packages/grafana-llm-frontend/src/mcp.tsx

```diff
@@ -0,0 +1,261 @@
+  let status: 'pending' | 'success' | 'error' = 'pending';
+  let result: Client | null = null;
+  let error: Error | null = null;
+
+  const key = clientKey(appName, appVersion);
+  const promise = (async () => {
+    if (clientMap.has(key)) {
+      result = clientMap.get(key)!;
+      status = 'success';
+      return result;
+    }
+
+    try {
+      const client = new Client({
+        name: appName,
+        version: appVersion,
+      });
+      const transport = new GrafanaLiveTransport();
+      await client.connect(transport);
+      clientMap.set(key, client);
+      status = 'success';
+      result = client;
+      return client;
+    } catch (e) {
+      status = 'error';
+      error = e as Error;
+      throw e;
+    }
+  })();
+
+  return {
+    read() {
+      if (status === 'pending') {
+        throw promise;
+      } else if (status === 'error') {
+        throw error;
+      } else if (status === 'success' && result) {
+        return result;
+      }
+      throw new Error('Unexpected resource state');
+    },
+  };
+}
+
+interface MCPClientProviderProps {
+  appName: string;
+  appVersion: string;
+  children: React.ReactNode;
+}
+
+/**
+ * MCPClientProvider is a React context provider that creates an MCP client
+ * and manages its lifecycle.
+ *
+ * It should be used to wrap the entire application in a single provider.
+ * This ensures that the client is created once and reused across the application.
+ *
+ * It also supports Suspense, which will suspend the component until the client
+ * is ready. This allows you to use the client in components that are not yet
+ * ready, such as those that are loading data.
+ *
+ * Example usage:
+ * ```tsx
+ * <Suspense fallback={<LoadingPlaceholder />}>
+ *   <ErrorBoundary>
+ *     {({ error }) => {
+ *       if (error) {
+ *         return <div>Something went wrong: {error.message}</div>;
+ *       }
+ *       return (
+ *         <MCPClientProvider appName="MyApp" appVersion="1.0.0">
+ *           <YourComponent />
+ *         </MCPClientProvider>
+ *       );
+ *     }}
+ *   </ErrorBoundary>
+ * </Suspense>
+ * ```
+ */
+export function MCPClientProvider({
+  appName,
+  appVersion,
+  children,
+}: MCPClientProviderProps) {
+  const resource = createClientResource(appName, appVersion);
+
+  // This will either return the client or throw a promise/error.
+  // If it throws a promise, Suspense will suspend the component until it resolves.
+  // If it throws an error, it should be caught by an ErrorBoundary.
+  const client = resource.read();
+
+  // Cleanup when the component unmounts.
+  React.useEffect(() => {
+    return () => {
+      if (client) {
+        client.close();
+      }
+      console.log('deleting client');
+      clientMap.delete(clientKey(appName, appVersion));
+    };
+  }, [client, appName, appVersion]);
+
+  return (
+    <MCPClientContext.Provider value={client}>
+      {children}
+    </MCPClientContext.Provider>
+  );
+}
```

Comment:
Question for frontend folks especially: do we think this is the right abstraction? We don't _need_ to provide a provider, people can just do

```ts
const client = new Client({ appName, appVersion });
await client.connect(new GrafanaLiveTransport());
```

but the provider and hook feels quite nice to avoid too much prop drilling, avoid `useAsync`, and make use of `Suspense`.

cc @manderson-dev @obetomuniz and possibly @SandersAaronD

Reply:
I'd recommend sticking with the provider pattern. It slightly reduces flexibility but significantly cuts down on boilerplate, avoids prop drilling, and discourages inconsistency across the codebase. It also allows making use of Suspense/ErrorBoundary. 

Still, if I understood correctly, if a specific need for flexibility arises, I think the `GrafanaLiveTransport` export covers that case neatly. Thoughts?

Reply:
Thanks @obetomuniz! The only concern I have is that the client will be shared between different apps and versions but I can't think of many situations where that would be a real issue, so let's stick with it.